File: apps/zxtune-android/src/app/zxtune/sound/WaveWriteSamplesTarget.java
/**
 * 
 * @file
 *
 * @brief
 *
 * @author vitamin.caig@gmail.com
 * 
 */

package app.zxtune.sound;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;

public final class WaveWriteSamplesTarget implements SamplesTarget {
  
  static final int SAMPLERATE = 44100;
  static final int CHANNELS = 1;
  static final int BYTES_PER_SAMPLE = 2;
  
  final byte header[] = {
      //+0
      'R', 'I', 'F', 'F',
      //+4 - file size - 8
      0, 0, 0, 0,
      //+8
      'W', 'A', 'V', 'E',
      //+12 - chunkId
      'f', 'm', 't', ' ',
      //+16 - chunkSize=16
      16, 0, 0, 0,
      //+20 - compression
      1, 0,
      //+22 - channels
      CHANNELS, 0,
      //+24 - samplerate
      0, 0, 0, 0,
      //+28 - bytes per sec
      0, 0, 0, 0,
      //+32 - align
      BYTES_PER_SAMPLE, 0,
      //+34 - bits per sample
      8 * BYTES_PER_SAMPLE, 0,
      //+36
      'd', 'a', 't', 'a',
      //+40 - data size
      0, 0, 0, 0
  };
  static final int OFFSET_WAVESIZE = 4;
  static final int OFFSET_SAMPLERATE = 24;
  static final int OFFSET_BPS = 28;
  static final int OFFSET_DATA_SIZE = 40;
  byte buffer[];
  final RandomAccessFile file;
  int doneBytes;

  public WaveWriteSamplesTarget(String filename) {
    this.buffer = new byte[CHANNELS * BYTES_PER_SAMPLE * SAMPLERATE];
    try {
      this.file = new RandomAccessFile(filename, "rw");
    } catch (FileNotFoundException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public int getSampleRate() {
    return SAMPLERATE;
  }

  @Override
  public int getPreferableBufferSize() {
    return SAMPLERATE;//1s buffer
  }

  @Override
  public void start() {
    try {
      file.seek(header.length);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    doneBytes = 0;
  }

  @Override
  public void writeSamples(short[] input) {
    final int inSamples = input.length / SamplesSource.Channels.COUNT;
    final int outBytes = inSamples * SamplesSource.Sample.BYTES; 
    allocateBuffer(outBytes);
    convertBuffer(input, outBytes);
    try {
      file.write(buffer, 0, outBytes);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    doneBytes += outBytes;
  }

  @Override
  public void stop() {
    setHeaderLE32(OFFSET_SAMPLERATE, SAMPLERATE);
    setHeaderLE32(OFFSET_BPS, SAMPLERATE * BYTES_PER_SAMPLE);
    setHeaderLE32(OFFSET_DATA_SIZE, doneBytes);
    setHeaderLE32(OFFSET_WAVESIZE, header.length - 8 + doneBytes);
    try {
      file.seek(0);
      file.write(header);
      file.seek(header.length + doneBytes);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public void release() {
  }
  
  private void allocateBuffer(int size) {
    if (buffer.length < size) {
      buffer = new byte[size];
    }
  }
  
  private void convertBuffer(short[] input, int outBytes) {
    assert BYTES_PER_SAMPLE == 2;
    assert CHANNELS == 1;
    for (int in = 0, out = 0; out < outBytes; )
    {
      int sample = 0;
      for (int ch = 0; ch != SamplesSource.Channels.COUNT; ++ch) {
        sample += input[in++];
      }
      sample /= SamplesSource.Channels.COUNT;
      buffer[out++] = (byte) (sample & 0xff);
      buffer[out++] = (byte) ((sample >> 8) & 0xff);
    }
  }
  
  private void setHeaderLE32(int offset, int value) {
    for (int b = 0; b != 4; ++b) {
      header[offset++] = (byte) (value & 0xff);
      value >>= 8;
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/sound/WaveWriteSamplesTarget.kt
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.sound

import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel

class WaveWriteSamplesTarget : SamplesTarget {

    companion object {
        private const val SAMPLERATE = 44100
        private const val CHANNELS = 1
        private const val BYTES_PER_SAMPLE = Short.SIZE_BYTES

        private const val EXTENSION = ".wav"

        private fun maybeAddExtension(filename: String) =
            if (filename.endsWith(EXTENSION, ignoreCase = true)) {
                filename
            } else {
                filename + EXTENSION
            }
    }

    private class Header {
        private val content = ByteBuffer.allocate(44).order(ByteOrder.LITTLE_ENDIAN)

        companion object {
            private const val OFFSET_WAVESIZE = 4
            private const val OFFSET_DATA_SIZE = 40
        }

        init {
            with(content) {
                // +0 - signature 'R', 'I', 'F', 'F'
                put(byteArrayOf(0x52, 0x49, 0x46, 0x46))
                // +4 - file size - 8
                putInt(0)
                //+8 - 'W', 'A', 'V', 'E', 'f', 'm', 't', ' '
                put(byteArrayOf(0x57, 0x41, 0x56, 0x45, 0x66, 0x6d, 0x74, 0x20))
                //+16 - chunkSize=16
                putInt(BYTES_PER_SAMPLE * 8)
                //+20 - compression
                putShort(1)
                //+22 - channels
                putShort(CHANNELS.toShort())
                //+24 - samplerate
                putInt(SAMPLERATE)
                //+28 - bytes per sec
                putInt(SAMPLERATE * BYTES_PER_SAMPLE)
                //+32 - align
                putShort(BYTES_PER_SAMPLE.toShort())
                //+34 - bits per sample
                putShort((8 * BYTES_PER_SAMPLE).toShort())
                //+36 - 'd', 'a', 't', 'a',
                put(byteArrayOf(0x64, 0x61, 0x74, 0x61))
                //+40 - data size
                putInt(0)
            }
        }

        val size
            get() = content.capacity()

        fun setDataSize(dataSize: Int): Unit = with(content) {
            position(OFFSET_DATA_SIZE)
            putInt(dataSize)
            position(OFFSET_WAVESIZE)
            putInt(size - 8 + dataSize)
        }

        fun writeTo(file: FileChannel): Unit = with(file) {
            val pos = position()
            position(0)
            write(content.apply { rewind() })
            position(pos)
        }
    }

    private val file: FileChannel
    private val header = Header()
    private lateinit var buffer: ByteBuffer
    private var doneBytes: Int = 0

    constructor(filename: String) : this(
        RandomAccessFile(maybeAddExtension(filename), "rw")
            .channel
    )

    constructor(channel: FileChannel) {
        file = channel
    }

    override val sampleRate = SAMPLERATE
    override val preferableBufferSize = SAMPLERATE //1s buffer

    override fun start() {
        file.position(header.size.toLong())
        doneBytes = 0
    }

    override fun writeSamples(buffer: ShortArray) {
        val inSamples = buffer.size / SamplesSource.Channels.COUNT
        val outBytes = inSamples * SamplesSource.Sample.BYTES
        allocateBuffer(outBytes)
        convertBuffer(buffer)
        file.write(this.buffer.apply { flip() })
        doneBytes += outBytes
    }

    override fun stop() = with(header) {
        setDataSize(doneBytes)
        writeTo(file)
    }

    override fun release() = file.close()

    private fun allocateBuffer(size: Int) {
        if (!this::buffer.isInitialized || buffer.capacity() < size) {
            buffer = ByteBuffer.allocate(size).order(ByteOrder.LITTLE_ENDIAN)
        }
    }

    private fun convertBuffer(input: ShortArray) = with(buffer) {
        clear()
        for (inp in input.indices step SamplesSource.Channels.COUNT) {
            var sample = 0
            for (ch in 0 until SamplesSource.Channels.COUNT) {
                sample += input[inp + ch]
            }
            buffer.putShort((sample / SamplesSource.Channels.COUNT).toShort())
        }
    }
}
